test/integration/certificates/readiness_controller_test.go

```diff
@@ -0,0 +1,185 @@
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificates"
+	"github.com/jetstack/cert-manager/pkg/controller/certificates/readiness"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/metrics"
+	"github.com/jetstack/cert-manager/test/integration/framework"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+)
+
+// TestReadinessController performs a basic test to ensure that
+// readiness controller sets the correct Ready condition on a Certificate.
+func TestReadinessController(t *testing.T) {
+	config, stopFn := framework.RunControlPlane(t)
+	defer stopFn()
+
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*20)
+	defer cancel()
+
+	// Fix the time to be able to test expired certificate scenario.
+	fakeClock := &fakeclock.FakeClock{}
+	// Build, instantiate and run the trigger controller.
+	kubeClient, factory, cmCl, cmFactory := framework.NewClients(t, config)
+
+	namespace := "testns"
+	certName := "test"
+	secretName := "test"
+	// notBefore value on the test certificate
+	notBefore := fakeClock.Now().UTC().Add(time.Hour).Truncate(time.Second)
+	// notAfter value on the test certificate
+	notAfter := notBefore.Add(time.Hour * 4).Truncate(time.Second)
+	// expected renewal time
+	renewalTime := notBefore.Add(time.Hour * 2).Truncate(time.Second)
+	issuer := cmmeta.ObjectReference{
+		Name:  "testissuer",
+		Kind:  "IssuerKind",
+		Group: "group.example.com",
+	}
+
+	// Create Namespace.
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+	_, err := kubeClient.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctrl, queue, mustSync := readiness.NewController(logf.Log, cmCl, factory, cmFactory, readiness.NewReadinessPolicyChain(fakeClock), certificates.RenewalTimeWrapper(cmapi.DefaultRenewBefore), readiness.PolicyEvaluator)
+	c := controllerpkg.NewController(
+		context.Background(),
+		"readiness_test",
+		metrics.New(logf.Log),
+		ctrl.ProcessItem,
+		mustSync,
+		nil,
+		queue,
+	)
+	stopController := framework.StartInformersAndController(t, factory, cmFactory, c)
+	defer stopController()
+
+	// 1. A newly created Certificate without a Secret should not be set to Ready.
+
+	// create the Certificate
+	cert, err := cmCl.CertmanagerV1().Certificates(namespace).Create(ctx,
+		gen.Certificate(certName,
+			gen.SetCertificateNamespace(namespace),
+			gen.SetCertificateSecretName(secretName),
+			gen.SetCertificateCommonName("example.com"),
+			gen.SetCertificateIssuer(issuer),
+			gen.SetCertificateRenewBefore(time.Hour*2)),
+		metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	// ensure that Certificate does not have Ready conditon set to True
+	ensureConditionNotApplied(ctx, t, cmCl, cert, cmapi.CertificateCondition{
+		Type:   cmapi.CertificateConditionReady,
+		Status: cmmeta.ConditionTrue,
+	})
```

Comment:
It seems like we are just waiting for a period of time to make sure it _doesn't_  exist. We should merge https://github.com/jetstack/cert-manager/pull/3613 first, and we can check here that another relevant condition is set for this spec generation.

Reply:
I guess we should observe that the readiness is set _False_ for the same generation.

Reply:
Good idea, I had not thought of this use case for the `observedGeneration` field!

> see if another relevant condition is set for this spec generation

Is `Ready=False` set right away (as opposed to only when an error occurs?)

If not, it seems like no condition will be set and it wouldn't be possible to use the observedGeneration trick 😞